app/src/main/java/mozilla/lockbox/store/AccountStore.kt

```diff
@@ -194,16 +203,56 @@ open class AccountStore(
             .subscribe(profileSubject::onNext, this::pushError)
             .addTo(compositeDisposable)
 
+        val token = securePreferences.getString(Constant.Key.accessToken)?.let {
+            accessTokenInfoFromJSON(JSONObject(it))
+        }
+
+        token?.let { handleAccessToken(it, isNew) } ?: tokenRotationHandler.post { fetchFreshToken(isNew) }
+    }
+
+    private fun fetchFreshToken(isNewLogin: Boolean = false) {
+        val fxa = fxa ?: return
         fxa.getAccessTokenAsync(Constant.FxA.oldSyncScope)
             .asMaybe(coroutineContext)
             .delay(1L, TimeUnit.SECONDS)
-            .map {
-                generateSyncCredentials(it, isNew).asOptional()
-            }
-            .subscribe(syncCredentialSubject::onNext, this::pushError)
+            .subscribe({ token ->
+                handleAccessToken(token, isNewLogin)
+            }, this::pushError)
             .addTo(compositeDisposable)
     }
 
+    private fun handleAccessToken(
+        token: AccessTokenInfo,
+        isNewLogin: Boolean
+    ) {
+        // We've just got a new token. It might be from secure preferences (we've just been
+        // restarted) or a token refresh.
+        // 1. Update the rest of the app.
+        val credentials = generateSyncCredentials(token, isNewLogin)
+        val syncCredentialSubject = syncCredentials as Subject
+        syncCredentialSubject.onNext(credentials.asOptional())
+
+        // 2. Store this token in the secure preferences.
+        securePreferences.putString(Constant.Key.accessToken, token.toJSONObject().toString())
+
+        // 3. Schedule a token refresh. It may be quite a long time away.
+        // Calculate how long before the token expires in milliseconds.
+        val msDelay = token.expiresAt * 1000L - timingSupport.currentTimeMillis
+
+        // We'll wait until it's almost expired, leaving at most 10 minutes to fetch the token.
```

Comment:
You don't need any of this scheduling stuff, really.

Next time you need a token, try to fetch it from your cache. If it's all good, nothing to do. If it's expired, fetch a new one from the library, cache it. That's all you need to do.

These tokens live 1-3 days, or so, so you'd only need to refresh them roughly once a day. The desire is to shorten their lifespan sometime in the future, but when that will happen is unclear, and even then it'll be on the order of a "a few hours". It's _very_ unlikely that your app won't be killed by then.

Reply:
> You don't need any of this scheduling stuff, really.
> 
> Next time you need a token, try to fetch it from your cache. If it's all good, nothing to do. If it's expired, fetch a new one from the library, cache it. That's all you need to do.

Yes this works, and would be my preference.

However, this is meant to be a small intervention. 

For right or wrong reasons, the token is derived from a-s at start up, and sent for use elsewhere. The token rotation scheduling is an unfortunate side effect of this previous decision.

Changing from "`AccountStore` gets from a-s and pushes `DataStore`" to "`DataStore` pulling from an a-s cache" model is a large amount of work. 

I would rather spend that time moving re-writing `DataStore` in terms of an `FxaManager`.